### ISSUE TITLE
Trivial false diagnostic assigning polymorphic numeric types to `&dyn Display`

### ISSUE THREAD

**Issue by emilyyyylime** [rust-lang/rust-analyzer#18637]
**rust-analyzer version**: `rust-analyzer 1.83.0 (90b35a6 2024-11-26)`
**rustc version**: `rustc 1.83.0 (90b35a623 2024-11-26)`

**editor or extension**: Reproduced in Vim and Helix

**code snippet to reproduce**: 
```rust
let _: &dyn std::fmt::Display = &1;
```

**rust-analyzer diagnostics output**:
```shell
$ rust-analyzer diagnostics src/main.rs 
processing crate: tmp, module: /home/emilylime/tmp/src/main.rs
Error RustcHardError("E0308") from LineCol { line: 1, col: 36 } to LineCol { line: 1, col: 38 }: expected &dyn Display, found &i32

diagnostic scan complete

Error: diagnostic error detected

Stack backtrace:
   0: <anyhow::Error>::msg::<&str>
   1: std::sys::backtrace::__rust_begin_short_backtrace::<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>
   2: <<std::thread::Builder>::spawn_unchecked_<<stdx::thread::Builder>::spawn<<rust_analyzer::cli::flags::Diagnostics>::run::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#0}, core::result::Result<(), anyhow::Error>>::{closure#1} as core::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
   3: std::sys::pal::unix::thread::Thread::new::thread_start
   4: <unknown>
   5: <unknown>
```

This replicates with floating point literals too, but not if a type specifier is included (i.e. `0i32` or `1.3f64`). #18133 seems possibly related.
